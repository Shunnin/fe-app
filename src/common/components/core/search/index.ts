export { SearchBox } from './search.component';
