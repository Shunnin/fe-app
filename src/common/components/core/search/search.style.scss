@import 'style/foundation/config';

.search-box {
  height: 48px;
  border-radius: $b-border-radius-md;
  background: $c-wh;
  box-shadow: 0px 1px 3px rgba(0, 0, 0, 0.1);
  margin-bottom: $b-space-2x;

  position: relative;
  display: flex;
  align-items: center;
  z-index: 1;

  &__icon {
    margin-left: $b-space;
    fill: $c-lb7;
  }

  &__input {
    flex: 1;
    margin-left: $b-space;
    height: 48px;
    width: 100%;

    border: none;
    box-sizing: border-box;
    font-size: $b-font-size;
    background-color: $c-wh;
    color: $c-blk;
    border-radius: $b-border-radius-md;

    &:focus {
      outline: none;
    }

    &::placeholder {
      color: $c-gr1;
    }
  }

  &__suggestions {
    position: absolute;
    display: flex;
    flex-direction: column;
    background: $c-wh;
    box-shadow: 0px 1px 3px rgba(0, 0, 0, 0.1);
    width: 98%;
    left: 1%;
    top: 54px;
    border-radius: $b-border-radius;
    overflow: hidden;

    & &-item {
      color: $c-gr1;
      text-decoration: none;
      padding: $b-space-s $b-space;
      display: block;
      text-align: left;
      border-bottom: 1px solid $c-gr3;
      font-size: $b-font-size;

      &:hover {
        background-color: $c-lb3;
      }
    }
  }
}
