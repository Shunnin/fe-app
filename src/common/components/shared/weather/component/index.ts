export { Temperature } from './temperature/temperature.component';
export { WeatherCard } from './weather-card/weather-card.component';
export { WeatherItem } from './weather-item/weather-item.component';
export * from './weather-icon';
