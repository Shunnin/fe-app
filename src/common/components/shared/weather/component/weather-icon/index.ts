export { WeatherIcon } from './weather-icon.component';
export * from './weather-icon.constant';
