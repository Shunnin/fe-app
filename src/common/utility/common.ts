export const getNextSevenDays = (): string[] => {
  const days = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
  const next7Days = [];

  // eslint-disable-next-line no-magic-numbers
  for (let i = 0; i < 7; i++) {
    // eslint-disable-next-line no-magic-numbers
    next7Days.push(days[new Date(Date.now() + (i + 1) * 24 * 60 * 60 * 1000).getDay()]);
  }

  return next7Days;
};
