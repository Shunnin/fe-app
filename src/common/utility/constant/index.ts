export * from './number';
export * from './common';
export * from './url';
