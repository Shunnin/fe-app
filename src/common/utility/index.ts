export * from './route';
export * from './constant';
export * from './promise';
export * from './temperature';
export * from './common';
export * from './url';
