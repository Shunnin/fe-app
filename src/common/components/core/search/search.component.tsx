import { FC, memo, useEffect, useRef, useState, useCallback, ChangeEvent } from 'react';
import { DebounceInput } from 'react-debounce-input';
import { slice, map, isEmpty, split } from 'lodash-es';

import { INT_ZERO, DEFAULT_DEBOUNCE_TIME, EMPTY_FUNC, IFunction } from '../../../utility';
import { SearchIcon } from '../icon';

import { useClickOutside } from '../../../hooks';

import './search.style';

const MAX_SUGGESTION_ITEMS = 6;

interface ISuggestionItemProps {
  label: string;
  onClickSuggestion: IFunction;
  setShowSuggestions: IFunction;
  suggestionRef: any;
}
const SuggestionItem: FC<ISuggestionItemProps> = ({ label, onClickSuggestion, setShowSuggestions, suggestionRef }) => {
  const handleClickSuggestItem = useCallback(() => {
    const value = split(label, ',')[INT_ZERO];

    onClickSuggestion(value);
    setShowSuggestions(false);
  }, [onClickSuggestion, label, suggestionRef]);

  return (
    <div className="search-box__suggestions-item clickable" onClick={handleClickSuggestItem}>
      {label}
    </div>
  );
};

interface ISuggestionListProps {
  suggestions: string[];
  maxSuggestItem: number;
  showSuggestions: boolean;
  setShowSuggestions: IFunction;
  onClickSuggestion: IFunction;
}

const SuggestionList: FC<ISuggestionListProps> = memo(
  ({ maxSuggestItem, showSuggestions, suggestions, setShowSuggestions, onClickSuggestion }) => {
    const suggestionRef = useRef(null);

    useClickOutside(suggestionRef, () => setShowSuggestions(false));

    if (!showSuggestions || isEmpty(suggestions)) {
      return <></>;
    }

    const suggestItems = slice(suggestions, INT_ZERO, maxSuggestItem);

    return (
      <div className="search-box__suggestions" ref={suggestionRef}>
        {map(suggestItems, (item: string, index: number) => (
          <SuggestionItem
            key={index}
            label={item}
            onClickSuggestion={onClickSuggestion}
            setShowSuggestions={setShowSuggestions}
            suggestionRef={suggestionRef}
          />
        ))}
      </div>
    );
  },
);
SuggestionList.displayName = 'SuggestionList';

interface ISearchBoxProps {
  placeholder?: string;
  suggestions: string[];
  maxSuggestItem?: number;
  onClickSearch?: IFunction;
  onClickSuggestion?: IFunction;
}

export const SearchBox: FC<ISearchBoxProps> = ({
  placeholder,
  suggestions,
  maxSuggestItem = MAX_SUGGESTION_ITEMS,
  onClickSearch = EMPTY_FUNC,
  onClickSuggestion = EMPTY_FUNC,
}) => {
  const [optionSuggestions, setOptionSuggestions] = useState<string[]>([]);
  const [showSuggestions, setShowSuggestions] = useState(false);

  const handleSearchInputChanged = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      const keyword = event.target.value;

      onClickSearch(keyword);
    },
    [onClickSearch],
  );

  useEffect(() => {
    const ableShowSuggestions = !isEmpty(suggestions);

    setShowSuggestions(ableShowSuggestions);
    setOptionSuggestions(suggestions);
  }, [suggestions]);

  return (
    <div className="search-box">
      <SearchIcon className="search-box__icon" />

      <DebounceInput
        className="search-box__input"
        debounceTimeout={DEFAULT_DEBOUNCE_TIME}
        onChange={handleSearchInputChanged}
        placeholder={placeholder}
      />

      <SuggestionList
        maxSuggestItem={maxSuggestItem}
        showSuggestions={showSuggestions}
        suggestions={optionSuggestions}
        setShowSuggestions={setShowSuggestions}
        onClickSuggestion={onClickSuggestion}
      />
    </div>
  );
};
