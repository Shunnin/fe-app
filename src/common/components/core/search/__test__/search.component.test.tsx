import { render, screen, fireEvent } from '@testing-library/react';

import { DEFAULT_DEBOUNCE_TIME } from '../../../../utility/constant';

import { SearchBox } from '../search.component';

describe('<SearchBox />', () => {
  jest.useFakeTimers();

  const onClickSearchMock = jest.fn();
  const onClickSuggestionMock = jest.fn();
  const suggestions = ['Hanoi, VN', 'Hue, VN', 'HCM, VN'];
  const props = {
    placeholder: 'Search location',
    suggestions: [],
    onClickSearch: onClickSearchMock,
    onClickSuggestion: onClickSuggestionMock,
  };

  it('renders and matches the snapshot with ignore onClickSearch and onClickSuggestion', () => {
    const { placeholder, suggestions } = props;
    const { container } = render(<SearchBox placeholder={placeholder} suggestions={suggestions} />);

    const searchBoxEl = container.firstChild;
    const searchBoxPlaceHolder = screen.getByPlaceholderText(props.placeholder);

    expect(searchBoxEl).toMatchSnapshot();
    expect(searchBoxPlaceHolder).toBeInTheDocument();
  });

  it('renders and matches the snapshot with default props', () => {
    const { container } = render(<SearchBox {...props} />);

    const searchBoxEl = container.firstChild;
    const searchBoxPlaceHolder = screen.getByPlaceholderText(props.placeholder);

    expect(searchBoxEl).toMatchSnapshot();
    expect(searchBoxPlaceHolder).toBeInTheDocument();
  });

  it('renders and matches the snapshot with given suggestion list', () => {
    const maxSuggestItem = 2;
    const { container } = render(<SearchBox {...props} maxSuggestItem={maxSuggestItem} suggestions={suggestions} />);

    const searchBoxEl = container.firstChild;
    const suggestionListEl = container.getElementsByClassName('search-box__suggestions-item');

    expect(searchBoxEl).toMatchSnapshot();
    expect(suggestionListEl.length).toBe(maxSuggestItem);
    expect(searchBoxEl.lastChild).toHaveClass('search-box__suggestions');
  });

  it('rerender and returns number suggestion item when given suggestion list', () => {
    const mockProps = { ...props, suggestions };
    const { rerender, container } = render(<SearchBox {...mockProps} maxSuggestItem={1} />);

    const suggestionListEl = container.getElementsByClassName('search-box__suggestions-item');

    expect(suggestionListEl.length).toBe(1);

    rerender(<SearchBox {...mockProps} maxSuggestItem={2} />);

    expect(suggestionListEl.length).toBe(2);
  });

  describe('handles click', () => {
    it('should call onClickSuggestion once when click in suggestion items', () => {
      const { container } = render(<SearchBox {...props} suggestions={suggestions} />);
      const suggestionItemEl = screen.getByText(suggestions[0]);
      const suggestionListEl = container.getElementsByClassName('search-box__suggestions');

      fireEvent.click(suggestionItemEl);

      expect(onClickSuggestionMock).toHaveBeenCalledTimes(1);
      expect(onClickSearchMock).not.toHaveBeenCalled();
      expect(suggestionListEl.length).toBe(0);
    });

    it('should close suggestion when click outside', () => {
      const { container } = render(<SearchBox {...props} suggestions={suggestions} />);
      const suggestionListEl = container.getElementsByClassName('search-box__suggestions');

      fireEvent.mouseDown(document.body);

      expect(onClickSuggestionMock).not.toHaveBeenCalled();
      expect(onClickSearchMock).not.toHaveBeenCalled();
      expect(suggestionListEl.length).toBe(0);
    });
  });

  describe('handles input search', () => {
    it('should call onClickSearchMock once when search item', () => {
      const { container } = render(<SearchBox {...props} suggestions={suggestions} />);
      const searchInputEl = screen.getByPlaceholderText(props.placeholder);
      const suggestionListEl = container.getElementsByClassName('search-box__suggestions');

      fireEvent.change(searchInputEl, { target: { value: '2020-05-24' } });

      expect(onClickSearchMock).not.toBeCalled();

      jest.advanceTimersByTime(DEFAULT_DEBOUNCE_TIME + 100);

      expect(onClickSearchMock).toHaveBeenCalledTimes(1);
      expect(onClickSuggestionMock).not.toHaveBeenCalled();
      expect(suggestionListEl.length).toBe(1);
    });
  });
});
