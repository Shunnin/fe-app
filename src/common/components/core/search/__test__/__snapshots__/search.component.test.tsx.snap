// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBox /> renders and matches the snapshot with default props 1`] = `
<div
  class="search-box"
>
  <reactcomponent
    class="search-box__icon"
  />
  <input
    class="search-box__input"
    placeholder="Search location"
    type="text"
    value=""
  />
</div>
`;

exports[`<SearchBox /> renders and matches the snapshot with given suggestion list 1`] = `
<div
  class="search-box"
>
  <reactcomponent
    class="search-box__icon"
  />
  <input
    class="search-box__input"
    placeholder="Search location"
    type="text"
    value=""
  />
  <div
    class="search-box__suggestions"
  >
    <div
      class="search-box__suggestions-item clickable"
    >
      Hanoi, VN
    </div>
    <div
      class="search-box__suggestions-item clickable"
    >
      Hue, VN
    </div>
  </div>
</div>
`;

exports[`<SearchBox /> renders and matches the snapshot with ignore onClickSearch and onClickSuggestion 1`] = `
<div
  class="search-box"
>
  <reactcomponent
    class="search-box__icon"
  />
  <input
    class="search-box__input"
    placeholder="Search location"
    type="text"
    value=""
  />
</div>
`;
